exporter/metric/prometheus/prometheus.go

```diff
@@ -0,0 +1,262 @@
+type Options struct {
+	// Registerer is the prometheus registerer to register
+	// metrics with. Use nil to specify the default registerer.
+	Registerer prometheus.Registerer
+
+	// Gatherer is the prometheus gatherer to gather
+	// metrics with. Use nil to specify the default gatherer.
+	Gatherer prometheus.Gatherer
+
+	// DefaultHistogramBuckets is the default histogram buckets
+	// to use. Use nil to specify the default histogram buckets.
+	DefaultHistogramBuckets []float64
+
+	// DefaultSummaryObjectives is the default summary objectives
+	// to use. Use nil to specify the default summary objectives.
+	DefaultSummaryObjectives map[float64]float64
+
+	// OnRegisterError defines a method to call to when registering
+	// a metric with the registerer fails. Use nil to specify
+	// to panic by default when registering fails.
+	OnRegisterError func(err error)
+}
+
+// NewExporter returns a new prometheus exporter for prometheus metrics.
+func NewExporter(opts Options) (*Exporter, error) {
+	if opts.Registerer == nil {
+		opts.Registerer = prometheus.DefaultRegisterer
+	} else {
+		// A specific registerer was set, check if it's a registry and if
+		// no gatherer was set, then use that as the gatherer
+		if reg, ok := opts.Registerer.(*prometheus.Registry); ok && opts.Gatherer == nil {
+			opts.Gatherer = reg
+		}
+	}
+	if opts.Gatherer == nil {
+		opts.Gatherer = prometheus.DefaultGatherer
+	}
+	if opts.OnRegisterError == nil {
+		opts.OnRegisterError = func(err error) {
+			panic(err)
+		}
+	}
+
+	// TODO: should we make a "PullController" ?
+	go func() {
+		_ = http.ListenAndServe(":22022", promhttp.HandlerFor(opts.Gatherer, promhttp.HandlerOpts{}))
+	}()
+
+	return &Exporter{
+		registerer:      opts.Registerer,
+		gatherer:        opts.Gatherer,
+		onRegisterError: opts.OnRegisterError,
+
+		counters: make(map[metricID]prometheus.Counter),
+		gauges:   make(map[metricID]prometheus.Gauge),
+
+		counterVecs: make(map[string]*prometheus.CounterVec),
+		gaugeVecs:   make(map[string]*prometheus.GaugeVec),
+	}, nil
+}
+
+// Export exports the provide metric record to prometheus.
+func (e *Exporter) Export(_ context.Context, checkpointSet export.CheckpointSet) error {
+	var aggError error
+	checkpointSet.ForEach(func(record export.Record) {
+		agg := record.Aggregator()
+		if sum, ok := agg.(aggregator.Sum); ok {
+			value, err := sum.Sum()
+			if err != nil {
+				aggError = err
+				// TODO: halp, where are we doing here?
+				return
+			}
+
+			c, err := e.getCounter(record)
+			if err != nil {
+				// TODO: log a warning here?
+				return
+			}
+
+			c.Add(value.AsFloat64())
+		}
+
+		if gauge, ok := agg.(aggregator.LastValue); ok {
+			lv, _, err := gauge.LastValue()
+			if err != nil {
+				aggError = err
+				// TODO: halp
+				return
+			}
+
+			g, err := e.getGauge(record)
+			if err != nil {
+				// TODO: log a warning here?
+				return
+			}
+			g.Set(lv.AsFloat64())
+		}
+	})
+
+	return aggError
```

Comment:
I think if we change the `ForEach` signature, we can just `return checkpointSet.ForEach(...)` in this method, it's error will propagate up to the Meter (via whatever controller we end up with).

Reply:
This could be helpful for prometheus exporter but others may need to do things after the `ForEach`. Do you think it is still useful to change this now?

Reply:
Not in this PR. The other exporters wouldn't simply return the call to ForEach.